crates/noirc_evaluator/src/ssa/opt/mem2reg.rs

```diff
@@ -1,378 +1,501 @@
+        block: BasicBlockId,
+        mut references: Block,
+    ) {
+        let instructions = function.dfg[block].instructions().to_vec();
+        let mut last_stores = BTreeMap::new();
 
-                    if has_common_successor {
-                        protected_allocations.insert(address);
-                    }
-                }
-                Instruction::Load { mut address } => {
-                    address = self.fetch_load_value_to_substitute(block_id, address);
-
-                    let found_last_value = self.find_load_to_substitute(
-                        block_id,
-                        address,
-                        dfg,
-                        instruction_id,
-                        loads_to_substitute,
-                        load_values_to_substitute_per_block,
-                    );
-                    if !found_last_value {
-                        // We want to protect allocations that do not have a load to substitute
-                        protected_allocations.insert(address);
-                        // We also want to check for allocations that share a value
-                        // with the one we are protecting.
-                        // This check prevents the pass from removing stores to a value that
-                        // is used by reference aliases in different blocks
-                        protected_allocations.insert(dfg.resolve(address));
-                    }
-                }
-                Instruction::Call { arguments, .. } => {
-                    for arg in arguments {
-                        if Self::value_is_from_allocation(*arg, dfg) {
-                            protected_allocations.insert(*arg);
-                        }
-                    }
-                }
-                _ => {
-                    // Nothing to do
-                }
-            }
+        for instruction in instructions {
+            self.analyze_instruction(function, &mut references, instruction, &mut last_stores);
         }
 
-        // Identify any arrays that are returned from this function
-        if let TerminatorInstruction::Return { return_values } = block.unwrap_terminator() {
-            for value in return_values {
-                if Self::value_is_from_allocation(*value, dfg) {
-                    protected_allocations.insert(*value);
-                }
-            }
-        }
+        self.handle_terminator(function, block, &mut references, &mut last_stores);
 
-        // Substitute load result values
-        for (result_value, new_value) in load_values_to_substitute_per_block {
-            let result_value = dfg.resolve(*result_value);
-            dfg.set_value_from_id(result_value, *new_value);
+        // If there's only 1 block in the function total, we can remove any remaining last stores
+        // as well. We can't do this if there are multiple blocks since subsequent blocks may
+        // reference these stores.
```

Comment:
Is this fine even if we have a function with 1 block that has remaining stores, but the remaining stores are to aliases passed into the function?

Reply:
Nevermind I see in the definition of `remove_stores_that_do_not_alias_parameters` you check against aliases. Could we add that to this comment as well?

Reply:
I think the function name is descriptive enough here such that we don't need to add to the comment.

Reply:
Yeah fair enough